package/kernel/mac80211/patches/ath/982-ath10k-qca9880-ar1a-archer-c7-wdr7500.patch

```diff
@@ -0,0 +1,916 @@
+-				&ar->normal_mode_fw);
+-	if (ret) {
+-		ath10k_err(ar, "Could not init core: %d\n", ret);
+-		goto err_power_down;
++	if (ar->bus_param.chip_id != 0x043200ff) {
++		ret = ath10k_hif_power_up(ar, ATH10K_FIRMWARE_MODE_NORMAL);
++		if (ret) {
++			ath10k_err(ar, "Could not init hif: %d\n", ret);
++			goto err_off;
++		}
++
++		ret = ath10k_core_start(ar, ATH10K_FIRMWARE_MODE_NORMAL,
++					&ar->normal_mode_fw);
++		if (ret) {
++			ath10k_err(ar, "Could not init core: %d\n", ret);
++			goto err_power_down;
++		}
+ 	}
+ 
+ 	param = ar->wmi.pdev_param->pmf_qos;
+--- a/drivers/net/wireless/ath/ath10k/pci.c
++++ b/drivers/net/wireless/ath/ath10k/pci.c
+@@ -54,7 +54,7 @@ static const struct pci_device_id ath10k
+ 	/* PCI-E QCA988X V2 (Ubiquiti branded) */
+ 	{ PCI_VDEVICE(UBIQUITI, QCA988X_2_0_DEVICE_ID_UBNT) },
+ 
+-	{ PCI_VDEVICE(ATHEROS, QCA988X_2_0_DEVICE_ID) }, /* PCI-E QCA988X V2 */
++	{ PCI_VDEVICE(ATHEROS, QCA988X_2_0_DEVICE_ID) }, /* PCI-E QCA986X, QCA988X V1/V2 */
+ 	{ PCI_VDEVICE(ATHEROS, QCA6164_2_1_DEVICE_ID) }, /* PCI-E QCA6164 V2.1 */
+ 	{ PCI_VDEVICE(ATHEROS, QCA6174_2_1_DEVICE_ID) }, /* PCI-E QCA6174 V2.1 */
+ 	{ PCI_VDEVICE(ATHEROS, QCA99X0_2_0_DEVICE_ID) }, /* PCI-E QCA99X0 V2 */
+@@ -66,10 +66,7 @@ static const struct pci_device_id ath10k
+ };
+ 
+ static const struct ath10k_pci_supp_chip ath10k_pci_supp_chips[] = {
+-	/* QCA988X pre 2.0 chips are not supported because they need some nasty
+-	 * hacks. ath10k doesn't have them and these devices crash horribly
+-	 * because of that.
+-	 */
++	{ QCA988X_2_0_DEVICE_ID, QCA988X_HW_1_0_CHIP_ID_REV },
+ 	{ QCA988X_2_0_DEVICE_ID_UBNT, QCA988X_HW_2_0_CHIP_ID_REV },
+ 	{ QCA988X_2_0_DEVICE_ID, QCA988X_HW_2_0_CHIP_ID_REV },
+ 
+@@ -629,6 +626,7 @@ static void ath10k_pci_sleep_sync(struct
+ static void ath10k_bus_pci_write32(struct ath10k *ar, u32 offset, u32 value)
+ {
+ 	struct ath10k_pci *ar_pci = ath10k_pci_priv(ar);
++	unsigned long irq_flags;
+ 	int ret;
+ 
+ 	if (unlikely(offset + sizeof(value) > ar_pci->mem_len)) {
+@@ -644,7 +642,12 @@ static void ath10k_bus_pci_write32(struc
+ 		return;
+ 	}
+ 
++	spin_lock_irqsave(&pciwar_lock, irq_flags);
++	ioread32(ar_pci->mem + offset + 4);
++	ioread32(ar_pci->mem + offset + 4);
++	ioread32(ar_pci->mem + offset + 4);
+ 	iowrite32(value, ar_pci->mem + offset);
++	spin_unlock_irqrestore(&pciwar_lock, irq_flags);
+ 	ath10k_pci_sleep(ar);
+ }
+ 
+@@ -3320,6 +3323,7 @@ int ath10k_pci_wait_for_target_init(stru
+ static int ath10k_pci_cold_reset(struct ath10k *ar)
+ {
+ 	u32 val;
++	int i;
+ 
+ 	ath10k_dbg(ar, ATH10K_DBG_BOOT, "boot cold reset\n");
+ 
+@@ -3339,13 +3343,25 @@ static int ath10k_pci_cold_reset(struct
+ 	 * for any immediate pcie register access and cause bus error,
+ 	 * add delay before any pcie access request to fix this issue.
+ 	 */
+-	msleep(20);
++	for (i = 0; i < ATH_PCI_RESET_WAIT_MAX; i++) {
++		if (ath10k_pci_reg_read32(ar, RTC_STATE_ADDRESS) &
++					  RTC_STATE_COLD_RESET_MASK) {
```

Comment:
this one

Reply:
this code was used originally until author of this commit changed it without any relevant info: https://patchwork.kernel.org/patch/6713001/
if it is found to cause problems on some chipset, a separate cold reset procedure can be made

Reply:
You are reverting a 5-year-old commit from qualcomm themselves without knowing what it does. This doesn't sound promising.

Reply:
using static generic linux delay here is what cause data bus error. and reading cold reset status from dedicated hw register is what works. i highly doubt a 5-year old commit is valid

Reply:
> using static generic linux delay here is what cause data bus error. and reading cold reset status from dedicated hw register is what works.

It's what works on qca9880v1 and v2 which are chips back at that time. Again it's not guaranteed that this commit doesn't break any new things now.

> i highly doubt a 5-year old commit is valid

You may need to bring this question upstream for qualcomm, who make all these chips and got hold of all those technical documentations, first?

Reply:
made a separate cold reset for v1 here.